native/ex_tokenizers/src/tokenizer.rs

```diff
@@ -639,26 +571,36 @@ pub fn tokenizer_train_from_files(
     // Current version of rust lib panics on retrainging with another trainer.
     // This leads to unpredicted nif behaviour.
     // Unwind can be removed after fixes https://github.com/huggingface/tokenizers/issues/525
+
     let result = panic::catch_unwind(|| {
-        let mut tokenizer: std::sync::RwLockWriteGuard<
-            '_,
-            tokenizers::TokenizerImpl<_, _, _, _, _>,
-        > = tokenizer.resource.0.write().unwrap();
+        let mut new_tokenizer = tokenizer.resource.0.clone();
+        let new_model = ExTokenizersModel::new(
+            tokenizer
+                .resource
+                .0
+                .get_model()
+                .resource
+                .0
+                .read()
+                .unwrap()
+                .clone(),
+        );
+        new_tokenizer.with_model(new_model);
```

Comment:
Do we need to make this operations with model explicitly?
AFAIU, `ExTokenizersModels` impls `Clone`, so cloning tokenizer will call this implimentation and just increase ResourceArc counter. Am i missing something?

Reply:
Exactly, the default behaviour would increase the refcount, so both old and new instance of the tokenizer would point to the same model, therefore training would affect both. To avoid that we make an actual copy of the model. Does that make sense?

Reply:
Yep, got it!